sstables/mirrored_storage.cc

```diff
@@ -0,0 +1,352 @@
+// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-
+/*
+ * Copyright (C) 2015-present ScyllaDB
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "mirrored_storage.hh"
+
+#include <boost/algorithm/string.hpp>
+
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/when_all.hh>
+#include <seastar/coroutine/all.hh>
+#include <seastar/coroutine/parallel_for_each.hh>
+#include <seastar/http/exception.hh>
+#include <seastar/util/file.hh>
+
+#include "sstables/sstables.hh"
+#include "sstables/sstables_manager.hh"
+#include "sstables/sstable_version.hh"
+#include "sstables/writer.hh"
+
+#include "db/system_keyspace.hh"
+
+#include "utils/memory_data_sink.hh"
+#include "utils/s3/client.hh"
+
+namespace {
+    static const auto status_creating = "creating";
+    static const auto status_sealed = "sealed";
+    static const auto status_removing = "removing";
+}
+
+namespace sstables {
+
+mirrored_storage::mirrored_storage(shared_ptr<s3::client> s3_client,
+                               sstring bucket,
+                               sstring dir)
+    : _s3_client{std::move(s3_client)}
+    , _bucket{std::move(bucket)}
+    , _location{dir}
+    , _dir{dir} // TODO: respect the state
+{}
+
+mirrored_storage::~mirrored_storage() = default;
+
+// run in async context
+void mirrored_storage::open(sstable& sst) {
+    _open_local(sst);
+    _open_remote(sst);
+}
+
+future<> mirrored_storage::_seal_local(const sstable& sst) {
```

Comment:
This is the copy of filesystem_storage::seal() isn't it? And the same is for `_seal_remote` -- it's the copy of s3_storage::seal(). Why copying the code here instead of keeping a reference to _storage_-s from mirrored_storage() and call those methods instead?